### ISSUE TITLE
Juntao - Rebalance transaction may revert due to price difference between Chainlink and Uniswap

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-05-USSD-judging#576]
Juntao

high

# Rebalance transaction may revert due to price difference between Chainlink and Uniswap

## Summary
Rebalance transaction may revert when protocol tries to sell collaterals and buy USSD, due to math error cause by the calculation of _amountToBuyLeftUSD_.

## Vulnerability Detail
If USSD price is lower than DAI, protocol will do rebalancing by selling collaterals for DAI and buying USSD. 
Protocol will first get [_collateralval_](https://github.com/sherlock-audit/2023-05-USSD/blob/main/ussd-contracts/contracts/USSDRebalancer.sol#L116) and then calculate [_amountToSellUnits_](https://github.com/sherlock-audit/2023-05-USSD/blob/main/ussd-contracts/contracts/USSDRebalancer.sol#L121) based on it.
```solidity
uint256 collateralval = IERC20Upgradeable(collateral[i].token).balanceOf(USSD) * 1e18 / (10**IERC20MetadataUpgradeable(collateral[i].token).decimals()) * collateral[i].oracle.getPriceUSD() / 1e18;
```
```solidity
uint256 amountToSellUnits = IERC20Upgradeable(collateral[i].token).balanceOf(USSD) * ((amountToBuyLeftUSD * 1e18 / collateralval) / 1e18) / 1e18;
```
After selling, protocol will try to update [_amountToBuyLeftUSD_](https://github.com/sherlock-audit/2023-05-USSD/blob/main/ussd-contracts/contracts/USSDRebalancer.sol#L123):
```solidity
amountToBuyLeftUSD -= (IERC20Upgradeable(baseAsset).balanceOf(USSD) - amountBefore);
```
However, protocol gets collateral price from Chainlink and sells for DAI on Uniswap, it possible that the DAI bought amount is larger than expected (due to the price diff between Chainlink and Uniswap), so _amountToBuyLeftUSD_ might be less than the DAI balance difference between before selling and after selling, meaning `amountToBuyLeftUSD < (IERC20Upgradeable(baseAsset).balanceOf(USSD) - amountBefore)`, the transaction will revert due to math error.
Let's assume:
1. _amountToBuyLeftUSD_ is **1000e18** (1000u);
2. The collateral (WBTC) price got from Chainlink is 20000u, and protocol holds 2 WBTC token (2e8), so _collateralval_ is **40000e18** (40000u);
3. _amountToSellUnits_ should be **5e6** (2.5% * 2e8);
4. Protocol sells 5e6 wei and gets **1000e18 + 1** DAI in return;
5. Protocol tries to update _amountToBuyLeftUSD_ and transaction reverts due to **1000e18 - (1000e18 + 1)**.

## Impact
Rebalance transaction will be reverted.

## Code Snippet
https://github.com/sherlock-audit/2023-05-USSD/blob/main/ussd-contracts/contracts/USSDRebalancer.sol#L123
https://github.com/sherlock-audit/2023-05-USSD/blob/main/ussd-contracts/contracts/USSDRebalancer.sol#L136

## Tool used

Manual Review

## Recommendation
If _collateralval_ is larger than _amountToBuyLeftUSD_, there is no need to update _amountToBuyLeftUSD_, please consider to remove [L123]([_amountToBuyLeftUSD_](https://github.com/sherlock-audit/2023-05-USSD/blob/main/ussd-contracts/contracts/USSDRebalancer.sol#L123));
If _collateralval_ is less than or equal to _amountToBuyLeftUSD_, please consider to update _amountToBuyLeftUSD_ as below:
```solidity
uint256 diff = (IERC20Upgradeable(baseAsset).balanceOf(USSD) - amountBefore);
if (amountToBuyLeftUSD >= diff)
    amountToBuyLeftUSD -= diff;
else
    amountToBuyLeftUSD = 0;
```

Duplicate of #111

**Comment by juntzhan** [comment]
Escalate for 10 USDC

This is a valid issue and it is well described in the report.
It's possbile (not seldom) that actual Uniswap swap price is larger than the price returned by Chainlink Price Feed, as they adopts different data source, and the risk is rebalancing would be reverted due to underflow error.

**Comment by hrishibhat** [comment]
@ctf-sec

**Comment by ctf-sec** [comment]
It is possible the rebalance revert

but I think if the collateral[i].path config it correctly, it will not revert

```solidity
 if (collateral[i].pathsell.length > 0) {
            uint256 amountBefore = IERC20Upgradeable(baseAsset).balanceOf(USSD);
            uint256 amountToSellUnits = IERC20Upgradeable(collateral[i].token).balanceOf(USSD) * ((amountToBuyLeftUSD * 1e18 / collateralval) / 1e18) / 1e18;
            IUSSD(USSD).UniV3SwapInput(collateral[i].pathsell, amountToSellUnits);
            amountToBuyLeftUSD -= (IERC20Upgradeable(baseAsset).balanceOf(USSD) - amountBefore);
            DAItosell += (IERC20Upgradeable(baseAsset).balanceOf(USSD) - amountBefore);
          }
```

the admin can config collateral[i].path [here](https://github.com/sherlock-audit/2023-05-USSD/blob/6d7a9fdfb1f1ed838632c25b6e1b01748d0bafda/ussd-contracts/contracts/USSD.sol#L84)

```solidity
    function addCollateral(
        address _address,
        address _oracle,
        bool _mint,
        bool _redeem,
        uint256[] calldata _ratios,
        bytes memory _pathbuy,
        bytes memory _pathsell,
        uint256 index
    ) public onlyControl {
        CollateralInfo memory newCollateral = CollateralInfo({
            token: _address,
            mint: _mint,
            redeem: _redeem,
            oracle: IStableOracle(_oracle),
            pathbuy: _pathbuy,
            pathsell: _pathsell,
            ratios: _ratios
        });
        if (index < collateral.length) {
            collateral[index] = newCollateral; // for editing
        } else {
            collateral.push(newCollateral); // for adding new collateral
        }
    }
```

**Comment by juntzhan** [comment]
1. Configuring collateral[i].path will not help because admin can't be sure when the config transaction will be executed and the configuration may not be correct by then;
2. Even if it does, admin can configure only after rebalance transaction was already reverted, because admin won't know the transaction will revert beforehand;
3. Configuring collateral[i].path is just a one-time fixing, the same issue still exists.

The rootcause is the price datasource is different between Chainlink and Uniswap Pool, no way to ensure the price will always be consistent.

**Comment by hrishibhat** [comment]
@juntzhan 
This seems like a valid duplicate of #111

**Comment by juntzhan** [comment]
> @juntzhan This seems like a valid duplicate of #111

Agree, thanks

**Comment by hrishibhat** [comment]
Result:
Medium
Duplicate of #111

**Comment by sherlock-admin** [comment]
Escalations have been resolved successfully!

Escalation status:
- [juntzhan](https://github.com/sherlock-audit/2023-05-USSD-judging/issues/576/#issuecomment-1605244157): accepted
